### ISSUE TITLE
Consider an alternative version of ShowSing that can be derived more easily

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/singletons#426]
When [`Sing` became a type family](https://github.com/goldfirere/singletons/issues/318), we had to surgically alter the `ShowSing` class to make it possible to mention `Sing` in the head of a quantified constraint:

```hs
class    (forall (z :: k). ShowSing' z) => ShowSing k
instance (forall (z :: k). ShowSing' z) => ShowSing k

class    Show (Sing z) => ShowSing' z
instance Show (Sing z) => ShowSing' z
```

This came at a cost, however. It is no longer possible to derive `Show` instances for `Sing` types like this:

```hs
deriving instance ShowSing k => Show (SBlah (z :: k))
```

Instead, one has to write out the full implementation of `showsPrec` like so in order to make it typecheck:

```hs
instance ShowSing k => Show (SBlah (z :: k)) where
  showsPrec p (SMkBlah (s :: Sing k)) =
    showParen (p > 10) $ showString "SMkBlah " . showsPrec 11 s
      :: ShowSing' k => ShowS
```

See [here](https://github.com/goldfirere/singletons/blob/013264115564b707f7cf440e97e9a3387dfb9987/src/Data/Singletons/ShowSing.hs#L116-L194) for an explanation. This is not the end of the world, since most `Show` instances for `Sing`s are autogenerated through Template Haskell anyway, but it does make implementing the few hand-written `Show` instances for `Sing`s much more annoying.

What if we could have our cake and eat it too? That is, what if we were able to tweak `ShowSing` so that it _could_ be used in standalone `deriving` declarations like the example I pose above? This may in fact be possible once https://gitlab.haskell.org/ghc/ghc/merge_requests/2283 lands in GHC HEAD, which tweaks how quantified constraints work to make it feasible for `singletons` to use the following version of `ShowSing`:

```hs
class    (forall (z :: k). ShowSing' z) => ShowSing k
instance (forall (z :: k). ShowSing' z) => ShowSing k

class    (forall (sing :: k -> Type). sing ~ Sing => Show (sing z))
                       => ShowSing' (z :: k)
instance Show (Sing z) => ShowSing' (z :: k)
```

As the patch for https://gitlab.haskell.org/ghc/ghc/merge_requests/2283 demonstrates, this allows the following code to typecheck:

```hs
newtype Foo a = MkFoo a
data SFoo :: forall a. Foo a -> Type where
  SMkFoo :: Sing x -> SFoo (MkFoo x)
type instance Sing = SFoo
deriving instance ShowSing a => Show (SFoo (z :: Foo a))

newtype Bar a = MkBar (Foo a)
data SBar :: forall a. Bar a -> Type where
  SMkBar :: forall a (x :: Foo a). Sing x -> SBar (MkBar x)
type instance Sing = SBar
deriving instance ShowSing (Foo a) => Show (SBar (z :: Bar a))
```

This is very exciting, as there is a very real possibility that this could work for everything in `singletons`. I decided to try this out myself to confirm this hunch.

Did it work? Yes! Well, mostly. There ~~were a couple of hiccups~~ was one hiccup along the way:

1. This instance:

   https://github.com/goldfirere/singletons/blob/013264115564b707f7cf440e97e9a3387dfb9987/src/Data/Singletons/ShowSing.hs#L274-L277

   No longer typechecks, even when I rewrite it to `deriving instance ShowSing k => Show (WrappedSing (a :: k))`:

    ```
   src/Data/Singletons/ShowSing.hs:277:1: error:
       • Couldn't match type ‘Sing’ with ‘WrappedSing’
           arising from a use of ‘GHC.Show.$dmshow’
       • In the expression: GHC.Show.$dmshow @(WrappedSing a)
         In an equation for ‘show’: show = GHC.Show.$dmshow @(WrappedSing a)
         When typechecking the code for ‘show’
           in a derived instance for ‘Show (WrappedSing a)’:
           To see the code I am typechecking, use -ddump-deriv
         In the instance declaration for ‘Show (WrappedSing a)’
       • Relevant bindings include
           show :: WrappedSing a -> String
             (bound at src/Data/Singletons/ShowSing.hs:277:1)
       |
   277 | deriving instance ShowSing k => Show (WrappedSing (a :: k))
       | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   
   src/Data/Singletons/ShowSing.hs:277:1: error:
       • Couldn't match type ‘Sing’ with ‘WrappedSing’
           arising from a use of ‘GHC.Show.$dmshowList’
       • In the expression: GHC.Show.$dmshowList @(WrappedSing a)
         In an equation for ‘showList’:
             showList = GHC.Show.$dmshowList @(WrappedSing a)
         When typechecking the code for ‘showList’
           in a derived instance for ‘Show (WrappedSing a)’:
           To see the code I am typechecking, use -ddump-deriv
         In the instance declaration for ‘Show (WrappedSing a)’
       • Relevant bindings include
           showList :: [WrappedSing a] -> ShowS
             (bound at src/Data/Singletons/ShowSing.hs:277:1)
       |
   277 | deriving instance ShowSing k => Show (WrappedSing (a :: k))
       | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   ```

   Very strange. I was able to make this typecheck again through the very painful process of inlining the default implementations of `show` and `showList`:

   ```hs
   instance ShowSing k => Show (WrappedSing (a :: k)) where
     showsPrec p (WrapSing s) = showParen (p >= 11) $ showsSing s
   
     show (WrapSing s) = showsSing s ""
   
     showList []              s = "[]" ++ s
     showList (WrapSing x:xs) s = '[' : showsSing x (showl xs)
       where
         showl []              = ']' : s
         showl (WrapSing y:ys) = showsSing y (showl ys)
   
   showsSing :: ShowSing k => Sing (x :: k) -> ShowS
   showsSing s = showString "WrapSing {unwrapSing = " . showsPrec 0 s . showChar '}'
   ```

   It feels very odd having to do this, however. I'm unclear if this is a GHC bug or a fundamental limitation of how quantified constraints work. (Note that like `Sing`, `WrappedSing` is also of kind `k -> Type`, which makes me wonder if that causes the local `(forall (sing :: k -> Type). sing ~ Sing => Show (sing z))` constraint to be fired too early.)
2. ~~I couldn't figure out how to make this instance typecheck at all, even with modifications:~~

   https://github.com/goldfirere/singletons/blob/013264115564b707f7cf440e97e9a3387dfb9987/src/Data/Singletons/Sigma.hs#L80-L86

   **EDIT**: Never mind, I was being silly here. I was trying to be too clever by changing `ShowSingApply'` to be more like `ShowSing'`.

Everything else Just Worked™. Still, the ~~two drawbacks~~ one drawback listed above makes me wonder if this idea is completely baked yet, so I opted to simply open an issue for discussing the idea before moving forward with a patch.

**Comment by goldfirere** [comment]
I will look at your comment above more closely in a moment. But in the meantime: would the plan in https://gitlab.haskell.org/ghc/ghc/issues/14860#note_163974 fix this whole mess? It looks like it would allow the desired quantified constraint that concludes a fact about `Show (Sing z)`.

**Comment by RyanGlScott** [comment]
If by "the plan" you mean "allow the heads of quantified constraints to mention quantified type variables as oversaturated arguments to type families", then yes, I think that would work just as well. Then we could just define `ShowSing` like this once and for all:

```hs
class    (forall (z :: k). Show (Sing z)) => ShowSing k
instance (forall (z :: k). Show (Sing z)) => ShowSing k
```

This ought to work since `type family Sing :: k -> Type`, so the `z` in `Show (Sing z)` is an oversaturated argument. Of course, I'm unclear on whether it's easier to fix GHC#14860 or diagnose the issue with `WrappedSing` above :)

**Comment by goldfirere** [comment]
Did you work on my branch from https://gitlab.haskell.org/ghc/ghc/merge_requests/2283? The instance in your comment is accepted on that branch.

**Comment by RyanGlScott** [comment]
Yes, I am using that branch. But I'm not sure what you mean by "The instance in your comment". I've tried the code in https://github.com/goldfirere/singletons/issues/426#issue-535351496 using your branch, but perhaps there's some subtlety I'm not conveying properly. Can you produce the exact code you compiled that you're referring to with your "instance in your comment" comment?

**Comment by goldfirere** [comment]
This is accepted on my branch:

```hs
{-# LANGUAGE QuantifiedConstraints, PolyKinds, TypeFamilies,
             GADTSyntax, FlexibleInstances, UndecidableInstances,
             FlexibleContexts, ScopedTypeVariables #-}

module T16502c where

import Data.Kind

type family Sing :: k -> Type

newtype WrappedSing :: forall k. k -> Type where
  WrapSing :: forall k (a :: k). { unwrapSing :: Sing a } -> WrappedSing a

class    (forall (z :: k). ShowSing' z) => ShowSing k
instance (forall (z :: k). ShowSing' z) => ShowSing k

class    Show (Sing z) => ShowSing' z
instance Show (Sing z) => ShowSing' z

instance ShowSing k => Show (WrappedSing (a :: k)) where
  showsPrec p (WrapSing s) = showParen (p >= 11) $
    showString "WrapSing {unwrapSing = " . showsPrec 0 s . showChar '}'
      :: ShowSing' a => ShowS
```

From your comment above, you seem to say that the instance "no longer type-checks". Maybe there's something else in the context that causes trouble, but the file above is accepted. (Sadly, there are other problems on my branch, as I've reported on the MR.)

**Comment by RyanGlScott** [comment]
Ah, there's the issue. You're using this version of `ShowSing'`:

```hs
class    Show (Sing z) => ShowSing' z
instance Show (Sing z) => ShowSing' z
```

As I alluded to earlier in https://github.com/goldfirere/singletons/issues/426#issue-535351496, this version doesn't work well with `StandaloneDeriving`. I'm instead using this version of `ShowSing'`:

```hs
class    (forall (sing :: k -> Type). sing ~ Sing => Show (sing z))
                       => ShowSing' (z :: k)
instance Show (Sing z) => ShowSing' (z :: k)
```

If you try this with `WrappedSing`, you will encounter trouble.

**Comment by RyanGlScott** [comment]
Unfortunately, it seems unlikely that GHC will ever support `deriving instance ShowSing k => Show (WrappedSing (a :: k))` using the version of `ShowSing'` defined in https://github.com/goldfirere/singletons/issues/426#issuecomment-564136649. See the discussion [here](https://gitlab.haskell.org/ghc/ghc/issues/16502#note_253431). I was able to at least make the hack I needed to employ to define a hand-written `Show` instance for `WrappedSing` slightly cleaner:

```hs
instance ShowSing k => Show (WrappedSing (a :: k)) where
  showsPrec = showsWrappedSingPrec
  show x = showsWrappedSingPrec 0 x ""
  showList = showListWith (showsWrappedSingPrec 0)

showsWrappedSingPrec :: ShowSing k => Int -> WrappedSing (a :: k) -> ShowS
showsWrappedSingPrec p (WrapSing s) = showParen (p >= 11) $
  showString "WrapSing {unwrapSing = " . showsPrec 0 s . showChar '}'
```

That's still rather sad, but it's a sight better than what I had in https://github.com/goldfirere/singletons/issues/426#issue-535351496.

I've checked in a working prototype in the [`T426`](https://github.com/goldfirere/singletons/tree/T426) branch. When `singletons` supports GHC 8.12 as the minimum, we can revisit whether we think this new version of `ShowSing'` is superior to the current one.
